### ISSUE TITLE
App uses the default cursor theme instead of the DE default

### ISSUE THREAD

**Issue by LiquidPL** [flathub/com.mikrotik.WinBox#9]
Here's how it looks in the Flatpak:

![image](https://github.com/user-attachments/assets/5278f27a-9eaa-48f0-8677-679ddbaececf)

Here's how it looks in the official release pulled from Mikrotik's website:

![image](https://github.com/user-attachments/assets/a391bb5a-a0a9-43e4-85f6-c0a79b8f28b5)

This is happening on Arch Linux, KDE Plasma 6.2.4.

**Comment by MatthieuCoder** [comment]
I was about to raise this issue. Seems like adding a XCURSOR_PATH environment variable makes winbox use the correct cursors.

**Comment by tomkv** [comment]
@LiquidPL 

Since I'm neither KDE nor Arch Linux user, please provide more info, how to replicate the issue.

What is the "DE default" cursor theme? You mean user configured theme? Or do you mean the theme that KDE ships with (Breeze)?

Please understand, that sandboxed applications normally do not have access outside the sandbox. That includes theme files on your host system, in `/usr/share/icons` or `~/.local/share/icons`. They can use only themes included inside their container, in this case, whatever comes with `org.freedesktop.Platform`. It does not come with Breeze, or other KDE themes.

@MatthieuCoder setting `XCURSOR_PATH` to which path? As per the note above, the sandboxed apps cannot access theme files outside sandbox. You need to make additional sandbox holes using a tool like flatpak cli or Flatseal. You can see a lengthy discussion on the topic in flatpak/flatpak#709.

**Comment by MatthieuCoder** [comment]
Hi, sorry for the late response,
Regarding the cursor issue, the initial resolution is based on my previous flatpak fork from @mkfrey

https://github.com/MatthieuCoder/com.mikrotik.WinBox/blob/main/com.mikrotik.WinBox.yaml

Setting the XCURSOR_PATH accordingly fixes the font issue on KDE Plasma (6.2.4) on Arch Linux
```
XCURSOR_PATH=/run/host/user-share/icons:/run/host/share/icons
```

I do not know enough about flatpak to explain *how* is fixes the issue, but adding it fixed the cursor issue for me.

**Comment by LiquidPL** [comment]
> What is the "DE default" cursor theme? You mean user configured theme? Or do you mean the theme that KDE ships with (Breeze)?

Yes, I mean Breeze in this case.

As for the sandbox issue, according to the link in https://github.com/flathub/com.mikrotik.WinBox/issues/9#issuecomment-2543988719 the fonts should be in the WinBox container, and after checking, they indeed are:

```
$ flatpak run --command=sh com.mikrotik.WinBox 
[📦 com.mikrotik.WinBox ~]$ ls -l /run/host/share/icons/
total 292
drwxr-xr-x 1 nfsnobody nfsnobody    144 Sep 28 17:36 Adwaita
drwxr-xr-x 1 nfsnobody nfsnobody    110 Sep 28 17:36 AdwaitaLegacy
drwxr-xr-x 1 nfsnobody nfsnobody    100 Dec  1 12:20 Breeze_Light
drwxr-xr-x 1 nfsnobody nfsnobody     36 Dec  1 12:20 KDE_Classic
drwxr-xr-x 1 nfsnobody nfsnobody     36 Dec  1 12:20 Oxygen_Black
drwxr-xr-x 1 nfsnobody nfsnobody     36 Dec  1 12:20 Oxygen_Blue
drwxr-xr-x 1 nfsnobody nfsnobody     36 Dec  1 12:20 Oxygen_White
drwxr-xr-x 1 nfsnobody nfsnobody     36 Dec  1 12:20 Oxygen_Yellow
drwxr-xr-x 1 nfsnobody nfsnobody     36 Dec  1 12:20 Oxygen_Zion
drwxr-xr-x 1 nfsnobody nfsnobody    266 Nov 27 11:46 breeze
drwxr-xr-x 1 nfsnobody nfsnobody    260 Nov 27 11:46 breeze-dark
drwxr-xr-x 1 nfsnobody nfsnobody    100 Dec  1 12:20 breeze_cursors
drwxr-xr-x 1 nfsnobody nfsnobody     22 Sep 28 17:36 default
drwxr-xr-x 1 nfsnobody nfsnobody    526 Dec 15 17:38 hicolor
drwxr-xr-x 1 nfsnobody nfsnobody     20 May 11  2024 locolor
drwxr-xr-x 1 nfsnobody nfsnobody     80 May 11  2024 oxygen
```

Also I checked this on a different flatpak I happened to have installed ([Prism Launcher](https://github.com/flathub/org.prismlauncher.PrismLauncher)), and over there the cursor theme is applied correctly to whatever I set in Plasma settings. Since I didn't find any reference to `/run/host/user-share/icons:/run/host/share/icons` in its yaml, I suppose the problem might lie in the difference between the KDE and fd.o platforms, and not this flatpak in particular.

**Comment by LiquidPL** [comment]
Yeah, the KDE runtime includes the host cursor paths by default: 

* https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/blob/master/elements/include/cursorpath.yml
* https://gitlab.com/freedesktop-sdk/freedesktop-sdk/-/blob/master/elements/components/xorg-lib-xcursor.bst#L16
* and various other references to `cursorpath`.

I suppose this is better to be handled in the fd.o runtime repository.

**Comment by MatthieuCoder** [comment]
Hi, 

Regarding the Prism Launcher Case it is based on the org.kde.Platform base & sdk.

**Comment by LiquidPL** [comment]
Yeah, that's what i mentioned at the end of my last comment. Something's broken in the freedesktop runtime that's not broken in the KDE runtime since cursors seem to work in that one. I've rebuilt this flatpak to use the KDE runtime and the cursors seem to work *better* (I think the winbox app just handles cursor changes badly but at least it will respect whatever is configured at the moment of launching it).

**Comment by tomkv** [comment]
Hello all,

so sofar, we know this:

- Winbox is Qt (not KDE) app. (X11-only one, at that; not relevant for this issue, but for the issue described bellow).
- Qt does respect KDE theming if it can. Which inside flatpak container might not be simple:
  - it has either hole in the sandbox to the host (as suggested above), because org.freedesktop.Platform doesn't come with DE-specific themes
  - or it is built against specific DE runtime
- but... there are other DEs than KDE, which also have cursor themes, with Qt plugins for integration and which usually work on non-sandboxed systems. But they won't work when using KDE runtime; the app will still have the same problem when used under non-KDE desktops.
- so given this, I'm not going into cursor theme topic, unless the runtime authors are going to solve this with their runtimes.

Then, there's an additional issue, related to cursors and tangentially to the cursor themes: cursor sizes. Currently, on hi-dpi xwayland systems, the cursor size is jumping between tiny and normal, as moved over the app window. I.e. in the right sidebar, it is tiny when over the menu items, but normal when it is over the blank space. For me personally, this is a more jarring than not matching theme, but similarly, something that has to be solved down  the stack.

So for the time being, I'm going to close this, but feel free to investigate and comment. If there is a good, non-hacky way to fix this for all DEs, I'm open to revisiting this topic. And of course, patches welcome ;-)

**Comment by MatthieuCoder** [comment]
The "hole" in the Sandbox is here by default, the only necessary "fix" that is necessary is adding an environment variable

**Comment by tomkv** [comment]
The default open paths are:

```
  - --filesystem=xdg-run/gvfs
  - --filesystem=/media
  - --filesystem=/run/media
  - --filesystem=/mnt
```

They are here because Winbox allows for browsing for the `Addresses.cdb` file, does not use portals and user-accessible areas -- homedir and mounted media -- are enough. It does not include `/run/host`, where the hack points with the env variable.

**Comment by MatthieuCoder** [comment]
> The default open paths are:
> 
> ```
>   - --filesystem=xdg-run/gvfs
>   - --filesystem=/media
>   - --filesystem=/run/media
>   - --filesystem=/mnt
> ```
> 
> They are here because Winbox allows for browsing for the `Addresses.cdb` file, does not use portals and user-accessible areas -- homedir and mounted media -- are enough. It does not include `/run/host`, where the hack points with the env variable.

No. The /run/host is there. It's included in flatpak since 0.8.8, it's not a hack. 
/https://docs.flatpak.org/en/latest/desktop-integration.html#icons
It's an actual functionality of flatpak.
